### PR TITLE
Password type validation method updates

### PR DESCRIPTION
**Pull request by mxstbr**
~~**WIP DO NOT MERGE YET**~~

Updates to the password field validation methods. I'm not sure how `validateRequiredInput` should behave, and if it behaves correctly at the moment – any comments @JedWatson?

### CODE REVIEW

File: fields/types/password/PasswordType.js [hunk 1]
```diff
@@ -123,10 +123,17 @@ password.prototype.validateInput = function (data, callback) {
...
 	var confirmValue = this.getValueFromData(data, '_confirm');
 	var passwordValue = this.getValueFromData(data);
-	if (passwordValue !== undefined && confirmValue !== undefined && passwordValue !== confirmValue) {
+	if (passwordValue !== undefined
+		&& confirmValue !== undefined
+		&& passwordValue !== confirmValue) {
 		result = false;
 		detail = 'passwords must match';
 	}
+	if ((passwordValue === undefined || confirmValue === undefined)
```
Comment (JedWatson) [inline fields/types/password/PasswordType.js]: This isn't necessary - if the `confirmValue` isn't provided, we don't validate it server-side. Password fields may be updated programatically or after client-side verification,

**Comment by JedWatson** [comment]
I worked on `validateRequiredInput` for Password fields recently (last week? the week before? recently.) and it's working as expected

File: fields/types/password/test/type.js [hunk 2]
```diff
@@ -8,6 +8,376 @@ exports.initList = function (List) {
...
+			var testItem = new List.model({
+				password: 'asdf',
+			});
+			List.fields.password.updateItem(testItem, {
+				password: null,
+			}, function () {
+				demand(testItem.password).be.null();
+				done();
+			});
+		});
```
Comment (mxstbr) [inline fields/types/password/test/type.js]: Should this be the case?

File: fields/types/password/test/type.js [hunk 3]
```diff
@@ -8,6 +8,376 @@ exports.initList = function (List) {
...
+			var testItem = new List.model({
+				password: 'asdf',
+			});
+			List.fields.password.updateItem(testItem, {
+				password: '',
+			}, function () {
+				demand(testItem.password).be('');
+				done();
+			});
+		});
```
Comment (mxstbr) [inline fields/types/password/test/type.js]: Should this be the case?

File: fields/types/password/test/type.js (hunk 2 above)
Comment (JedWatson) [inline fields/types/password/test/type.js]: Yep, that should clear the stored value. We should actually query and assert the stored value after a save to ensure we're not saving a hash of `null` (if that's even a thing)

File: fields/types/password/test/type.js (hunk 3 above)
Comment (JedWatson) [inline fields/types/password/test/type.js]: Yes, again with an assertion that the stored value is cleared as well.

File: fields/types/password/PasswordType.js (hunk 1 above)
Comment (mxstbr) [inline fields/types/password/PasswordType.js]: Alright. Do we check that a `password` was passed? Or if that's undefined but the `confirm` exists should it still pass?

Comment (JedWatson) [inline fields/types/password/PasswordType.js]: Good catch. If the `password` is undefined but a `confirmValue` was passed, I think we should fail validation as if they were both provided and didn't match.

Comment (mxstbr) [inline fields/types/password/PasswordType.js]: Alright, what about `password` is `null` or `""` and `confirmValue` is valid?

**Comment by mxstbr** [comment]
Updated, should now hit on all points. I don't yet check the stored value, but we'll do that in another PR

**Pull request merged**